### PR TITLE
8314745: JFR: @StackFilter

### PR DESCRIPTION
**Pull request by egahlin**
Could I have a review of a change that allows stack frames to be filtered out.

For now, the mechanism is only available for JDK events, but plan is to make the annotation public once we feel confident the API is correct.

Testing: tier1 + tier2 + jdk/jdk/jfr/

Thanks
Erik

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8314745](https://bugs.openjdk.org/browse/JDK-8314745): JFR: @StackFilter (**Enhancement** - P4)
 * [JDK-8315236](https://bugs.openjdk.org/browse/JDK-8315236): JFR: @StackFilter (**CSR**) (Withdrawn)

### Reviewers
 * [Markus Grönlund](https://openjdk.org/census#mgronlun) (@mgronlun - **Reviewer**) ⚠️ Review applies to [3e5c8e56](https://git.openjdk.org/jdk/pull/16726/files/3e5c8e56ef9ffc59494a995dd7e72ad1bfca2895)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16726/head:pull/16726` \
`$ git checkout pull/16726`

Update a local copy of the PR: \
`$ git checkout pull/16726` \
`$ git pull https://git.openjdk.org/jdk.git pull/16726/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16726`

View PR using the GUI difftool: \
`$ git pr show -t 16726`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16726.diff">https://git.openjdk.org/jdk/pull/16726.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16726#issuecomment-1817962281)

### CODE REVIEW

**Comment by AlanBateman** [comment]
I wonder about how maintainable this is. I assume it would be easy for the code to be refactored and the frames listed in the StackFilter annotation value to get out of date. Are there any options to put an annotation on the methods to be filtered, kinda like what we do with `@Hidden`? I realise this might restrict it to JDK events in java.base initially but it seems more maintainable that having event having a method list that isn't checkable at compile time.

**Review by mgronlun — APPROVED** [review]
Looks very nice! Good work.

File: src/jdk.jfr/share/classes/jdk/jfr/internal/EventControl.java [hunk 1]
```diff
@@ -89,18 +93,92 @@ record NamedControl(String name, Control control) {
...
+    private void addStackFilters(PlatformEventType eventType) {
+        String[] filter = getStackFilter(eventType);
+        if (filter != null) {
+            int size = filter.length;
+            List<String> types = new ArrayList<>(size);
+            List<String> methods = new ArrayList<>(size);
+            for (String frame : filter) {
+                int index = frame.indexOf("::");
+                String clazz = null;
+                String method  = null;
```
Comment (turbanoff) [inline src/jdk.jfr/share/classes/jdk/jfr/internal/EventControl.java:119]:
```suggestion
                String method = null;
```

**Pull request closed without merging**